geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/filters/Z3Filter.scala

```diff
@@ -8,23 +8,85 @@
 
 package org.locationtech.geomesa.index.filters
 
-import java.nio.ByteBuffer
-
-import org.locationtech.geomesa.index.filters.RowFilter.RowFilterFactory
+import com.typesafe.scalalogging.StrictLogging
+import org.geotools.util.factory.Hints
+import org.locationtech.geomesa.index.filters.RowFilter.{FilterResult, RowFilterFactory}
+import org.locationtech.geomesa.index.filters.Z3Filter.ZBounds
 import org.locationtech.geomesa.index.index.z3.Z3IndexValues
 import org.locationtech.geomesa.utils.index.ByteArrays
 import org.locationtech.sfcurve.zorder.Z3
 
+import java.nio.ByteBuffer
+
+/**
+ * Z3 filter implementation
+ *
+ * @param xy spatial bounds
+ *           <ul><li>each element of the outer array is a bounding box</li>
+ *           <li>the inner arrays wil always have 4 elements, consisting of (xmin,ymin,xmax,ymax)</li></ul>
+ *           for example, `bbox(0,1,2,3) OR bbox(4,5,6,7)` would be `Array(Array(0,1,2,3), Array(4,5,6,7))`
+ *
+ * @param t temporal bounds
+ *          <ul><li>each element of the outer array is the times for a particular epoch, indexed by (epoch - epochOffset)
+ *            if the epoch is < epochOffset or > t.length + epochOffset, or the element corresponding to the epoch is null, then
+ *            the entire epoch is included in the time filter</li>
+ *          <li>each element of the second array is a time bound within the given epoch</li>
+ *          <li>the inner arrays will always have 2 elements, consisting of (min time, max time)</li>
+ *          <li>the outer array may be empty, which indicates all epochs are fully included in the time filter</li></ul>
+ *          for example, `during t1/t2 OR during t3/t4` would be `Array(Array(Array(tz1,tz2), Array(tz3,tz4)))`
+ *            (assuming a single epoch). tz refers to the time offset within the epoch
+ *
+ * @param epochOffset offset into the temporal bounds array
+ * @param maxTime max time for our space-filling curve
+ * @param seek whether to seek or not
+ */
 class Z3Filter(
     val xy: Array[Array[Int]],
     val t: Array[Array[Array[Int]]],
-    val minEpoch: Short,
-    val maxEpoch: Short
-  ) extends RowFilter {
+    val epochOffset: Short,
+    val maxTime: Int,
+    val seek: Boolean
+  ) extends RowFilter with StrictLogging {
+
+  private val zBoundsNoTime: ZBounds = if (!seek) { null } else {
+    val bounds = xy.map { case Array(xmin, ymin, xmax, ymax) =>
+      Z3Filter.zBound(xmin, ymin, 0, xmax, ymax, maxTime)
```

Comment:
In the backwards compatibility case, maxTime is 0, right?

Reply:
seek will be false in that case so this line won't get executed (it's checked in the deserialize method).